monty-31/src/aarch64_neon/packing.rs

```diff
@@ -586,38 +617,229 @@ impl_packed_field_pow_2!(
+        );
+        sum01_h = aarch64::vmlal_high_u32(sum01_h, lhs[1].into_vec(), rhs[1].into_vec());
+
+        // Compute sum23 = (lhs[2] * rhs[2]) + (lhs[3] * rhs[3])
+        let mut sum23_l = aarch64::vmull_u32(
+            aarch64::vget_low_u32(lhs[2].into_vec()),
+            aarch64::vget_low_u32(rhs[2].into_vec()),
+        );
+        let mut sum23_h = aarch64::vmull_high_u32(lhs[2].into_vec(), rhs[2].into_vec());
+        sum23_l = aarch64::vmlal_u32(
+            sum23_l,
+            aarch64::vget_low_u32(lhs[3].into_vec()),
+            aarch64::vget_low_u32(rhs[3].into_vec()),
+        );
+        sum23_h = aarch64::vmlal_high_u32(sum23_h, lhs[3].into_vec(), rhs[3].into_vec());
+
+        // Combine the partial sums
+        let sum_l = aarch64::vaddq_u64(sum01_l, sum23_l);
+        let sum_h = aarch64::vaddq_u64(sum01_h, sum23_h);
+
+        // Split C into 32-bit halves per lane:
+        // - c_lo = C mod 2^{32},
+        // - c_hi = C >> 32.
+        let c_lo = aarch64::vuzp1q_u32(
+            aarch64::vreinterpretq_u32_u64(sum_l),
+            aarch64::vreinterpretq_u32_u64(sum_h),
+        );
+        let c_hi = aarch64::vuzp2q_u32(
+            aarch64::vreinterpretq_u32_u64(sum_l),
+            aarch64::vreinterpretq_u32_u64(sum_h),
+        );
+
+        // Partial high-word reduction: c_hi' ∈ [0,P).
+        //
+        // Since C < 4P^2 and P < 2^{31}, we have c_hi < 2P.
+        //
+        // Map [P,2P) → [0,P) by subtracting P.
+        let c_hi_sub = aarch64::vsubq_u32(c_hi, P::PACKED_P);
+        let c_hi_prime = aarch64::vminq_u32(c_hi, c_hi_sub);
+
+        // q ≡ c_lo ⋅ μ (mod 2^{32}), with μ = −P^{-1} (mod 2^{32}).
+        let q = aarch64::vmulq_s32(aarch64::vreinterpretq_s32_u32(c_lo), P::PACKED_MU);
+        let q_u32 = aarch64::vreinterpretq_u32_s32(q);
+
+        // Compute (q⋅P)_hi = high 32 bits of q⋅P per lane (exact unsigned widening multiply).
+        let qp_l = aarch64::vmull_u32(
+            aarch64::vget_low_u32(q_u32),
+            aarch64::vget_low_u32(P::PACKED_P),
+        );
+        let qp_h = aarch64::vmull_high_u32(q_u32, P::PACKED_P);
+        let qp_hi = aarch64::vuzp2q_u32(
+            aarch64::vreinterpretq_u32_u64(qp_l),
+            aarch64::vreinterpretq_u32_u64(qp_h),
+        );
+
+        let d = aarch64::vsubq_s32(
+            aarch64::vreinterpretq_s32_u32(c_hi_prime),
+            aarch64::vreinterpretq_s32_u32(qp_hi),
+        );
+
+        // Canonicalize d from (-P, P) to [0, P) branchlessly.
+        //
+        // The `vmlsq_u32` instruction computes `a - (b * c)`.
+        // - If `d` is negative, the mask is `-1` (all 1s), so we compute `d - (-1 * P) = d + P`.
+        // - If `d` is non-negative, the mask is `0`, so we compute `d - (0 * P) = d`.
+        let underflow = aarch64::vcltq_u32(c_hi_prime, qp_hi);
+        let canonical_res =
+            aarch64::vmlsq_u32(aarch64::vreinterpretq_u32_s32(d), underflow, P::PACKED_P);
```

Comment:
`d` is a u32 after the above change

```suggestion
        let canonical_res = aarch64::vmlsq_u32(d, underflow, P::PACKED_P);
```

Reply:
fixed